### ISSUE TITLE
Optimizing

### ISSUE THREAD

**Issue by GrilledCh3z** [OmniKraftDevelopers/CleanShop#1]
Hey @PeculiarCarrot just wanted to put this out here for when ever you get the chance. I was wanting to see if you're free to get on teamspeak sometime so we can go over optimizing a few things. I noticed a few things that we can do to help optimize.

**Comment by PeculiarCarrot** [comment]
I should be able to sometime tonight. By the way, you should probably know that I'm recreating the search shops function, since nobody seems to use it. The plugin will search chests of shops and add them to that shop's "stock". Players can then use /searchshops <item> to be given a list of shops that have the item. There's more depth to the design than that, but that's basically what I'm doing right now. Making good progress.

**Comment by GrilledCh3z** [comment]
Ah okay. You may want to be careful doing that though as it could be very cpu intensive and also taxing on the memory

**Comment by PeculiarCarrot** [comment]
Yeah, I realize that. It takes a lot of time to loop through 3D space. Right now I'm handling it by checking if a chest is in a shop when a player closes it. If so, I just recalculate the stock in that shop.

**Comment by GrilledCh3z** [comment]
That would be slightly better. However very abuse-able and also you would have to make sure every chest gets opened. Maybe have a command say /updatechest and the shop owner stands and looks at the chest and when they run the command it updates the chest they're looking at.

**Comment by PeculiarCarrot** [comment]
That won't work. People are too lazy to do that. I already have a command to add items/remove items, but nobody uses it. Also, I have TeamSpeak set up and ready, if you want talk.

**Comment by GrilledCh3z** [comment]
Yeah, but doesnt your command run based on each individual item? This would remove both of those commands and make it so that they could run one command in-order to update them.

**Comment by PeculiarCarrot** [comment]
I honestly don't think people would take the time. They'd forget, or just be like "nah". Besides, I tested this with a pretty large region, and got absolutely no lag.

**Comment by GrilledCh3z** [comment]
You have to consider though that we have 40+ shops. And how often would that updating run? 
Also in my opinion... The people that wont take 10-20 seconds to update their shop probably shouldn't get the ability to search the items :stuck_out_tongue_winking_eye:

**Comment by PeculiarCarrot** [comment]
I literally just ran it with 1.5 million blocks in survival mode and punched a slime block. The drop appeared instantly.

**Comment by GrilledCh3z** [comment]
It's not the size of the region i'm worried about. You have a very good chance of creating a memory leak with creating that many objects. But so how often would it update the contents?

**Comment by PeculiarCarrot** [comment]
Whenever a chest is closed as of right now. How do you even create a memory leak in Java?

**Comment by GrilledCh3z** [comment]
Hmm, that could be even more of an issue. That could be very easy to abuse. And memory leaks are basically when you create an instance of many objects that basically ends up taking a lot of memory out.

**Comment by PeculiarCarrot** [comment]
Alright, here's an idea. Why don't I just keep a list of the items a chest contains, and it's only updated when that chest is closed? That would take basically no cpu power, and very little memory. I don't know why I didn't do this originally...

**Comment by GrilledCh3z** [comment]
That could work. However again that could take up quite a bit of memory. You would have to use a hashmap with an arraylist inside. - at least that would be the most efficient way. And even then, if one of those was made for every chest in a shop, that could easily take over a gig of memory. And also garbage collection may screw that up because it could delete them if they arent used or referred back to

**Comment by PeculiarCarrot** [comment]
How would that be a gig of memory? 27 slots(bytes) in each chest, 30 chests per shop (that's way too many), in 50 shops is 40,500 bytes, or 40.5 kb.

**Comment by GrilledCh3z** [comment]
er.. A chest isnt 27 bytes... Its usually about 40 kilobytes in the memory. Every item inside the chest is it's own object and then the chest it's self is an object they all take up a LOT LOT more than 40.5 kb lol

**Comment by PeculiarCarrot** [comment]
But all I need to store is the ID of each item. I don't need every chest object or every ItemStack.

**Comment by GrilledCh3z** [comment]
You also have to consider though that it's not just item ids. It's that item id's data as well

**Comment by GrilledCh3z** [comment]
Are you writing them to a document? Or storing them in the memory like an arraylist?

**Comment by PeculiarCarrot** [comment]
I would store them in an arraylist most likely.

**Comment by PeculiarCarrot** [comment]
Also, I haven't been messing with data, mostly because nobody would use it. /searchshops stone 2? Nah. Just /searchshops stone, and look for the thing you want.

**Comment by GrilledCh3z** [comment]
The issue with storing it in an arraylist is that when the server reboots... all the info on what shops carry would be deleted

and you really should do date, because things like wool for example, if you want gray wool instead of white wool, they have different datas

**Comment by PeculiarCarrot** [comment]
I'd store it in a file, too. I thought that was a given. And regarding data, I'd have to hand-code every item with every piece of data and name it. I... really don't want to do that.

**Comment by GrilledCh3z** [comment]
no you wouldnt have to do that. I can really quickly write you a serialization method. It will turn the items into a string. and then you can convert them back into an item from that string. It will store everything really efficiently. Gimmie just a few

**Comment by GrilledCh3z** [comment]
here you go https://gist.github.com/killerkoda13/a423b9050c928191a8ef1bb0f3bacff7
That code will turn the items into a string encoded in base 64. That should make storing them really easy.

**Comment by PeculiarCarrot** [comment]
Erm... When I said I'd have to name them, I meant in order for players to search for the item in shops. Ex: /searchshops blue_wool

**Comment by GrilledCh3z** [comment]
In that cause you dont have to do that either. Have them search by material name. Spigot already has a method. Material.getMaterial(String name)

**Comment by GrilledCh3z** [comment]
But they dont have to specifically search for the names. They could still do WoolID:data
